poolakey/src/main/java/ir/cafebazaar/poolakey/BillingConnection.kt

```diff
@@ -62,38 +63,37 @@ internal class BillingConnection(
     }
 
     override fun onServiceConnected(name: ComponentName?, service: IBinder?) {
-        IInAppBillingService.Stub.asInterface(service)
-            ?.takeIf(
-                thisIsTrue = {
-                    isPurchaseTypeSupported(
-                        purchaseType = PurchaseType.IN_APP,
-                        inAppBillingService = it
-                    )
-                },
-                andIfNot = {
-                    callback?.connectionFailed?.invoke(IAPNotSupportedException())
-                }
-            )
-            ?.takeIf(
-                thisIsTrue = {
-                    !paymentConfiguration.shouldSupportSubscription || isPurchaseTypeSupported(
-                        purchaseType = PurchaseType.SUBSCRIPTION,
-                        inAppBillingService = it
-                    )
-                },
-                andIfNot = {
-                    callback?.connectionFailed?.invoke(SubsNotSupportedException())
-                }
-            )
-            ?.also { billingService = it }
-            ?.also { callback?.connectionSucceed?.invoke() }
+        try {
+            IInAppBillingService.Stub.asInterface(service)
+                ?.also { billingService = it }
+                ?.takeIf(
+                    thisIsTrue = {
+                        isPurchaseTypeSupported(purchaseType = PurchaseType.IN_APP)
+                    },
+                    andIfNot = {
+                        callback?.connectionFailed?.invoke(IAPNotSupportedException())
+                    }
+                )
+                ?.takeIf(
+                    thisIsTrue = {
+                        !paymentConfiguration.shouldSupportSubscription || isPurchaseTypeSupported(
+                            purchaseType = PurchaseType.SUBSCRIPTION
+                        )
+                    },
+                    andIfNot = {
+                        callback?.connectionFailed?.invoke(SubsNotSupportedException())
+                    }
+                )
+                ?.also { callback?.connectionSucceed?.invoke() }
+        } catch (ignored: DeadObjectException) {
+            // onServiceDisconnected will get called so no need to do anything
+        }
```

Comment:
We shouldn't suppress this exception. Developer should know that their app has disconnected from the IAB service. So, I'd suggest to do this instead:  
```kotlin
callback?.connectionFailed?.invoke(e)
```